File: src/main/java/org/kambanaria/knots/examples/ExtendedUserKnot.java
package org.kambanaria.knots.examples;

import java.util.List;
import org.kambanaria.knots.microservices_models.Account;
import org.kambanaria.knots.microservices_models.User;
import static org.kambanaria.knots.services.StaticRegistry.*;

public class ExtendedUserKnot {

    private final int userId;
    private User user;
    private List<Account> accounts;

    public ExtendedUserKnot(int userId) {
        this.userId = userId;
    }

    public int getUserId() {
        return userId;
    }

    public User getUser() {
        if (user == null) {
            user = USER_SERVICE.getUserById(userId);
        }
        return user;
    }

    public List<Account> getAccounts() {
        if (accounts == null) {
            accounts = ACCOUNT_SERVICE.getAccountByUserId(userId);
        }
        return accounts;
    }
}

File: src/main/python/extended_user_knot.py
from services import UserService as USER_SERVICE
from services import AccountService as ACCOUNT_SERVICE


class ExtendedUserKnot(object):

    def __init__(self, user_id):
        self.user_id = user_id
        self._cached_user = None
        self._cached_accounts = None

    @property
    def user(self):
        if self._cached_user is None:
            self._cached_user = USER_SERVICE.get_user_by_id(self.user_id)
        return self._cached_user

    @property
    def accounts(self):
        if self._cached_accounts is None:
            self._cached_accounts = ACCOUNT_SERVICE.get_account_by_user_id(self.user_id)
        return self._cached_accounts
